File: jt/StatusRenderer.java
package jt;

import javax.swing.*;
import javax.swing.table.TableCellRenderer;
import java.awt.*;

public class StatusRenderer extends JPanel implements TableCellRenderer
{
	public static final StatusRenderer INSTANCE = new StatusRenderer();

	public int height;

	JLabel icon;
	JTextArea name;
	JTextArea text;
	JTextArea via;

	public StatusRenderer()
	{
		this.setLayout(null);
		icon = new JLabel(new ImageIcon("a.png"));
		this.add(icon);
		icon.setSize(48, 48);
		icon.setLocation(5, 5);

		name = new JTextArea();
		name.setEditable(false);
		name.setLineWrap(true);
		this.add(name);

		text = new JTextArea();
		text.setEditable(false);
		text.setLineWrap(true);
		this.add(text);

		via = new JTextArea();
		via.setEditable(false);
		via.setLineWrap(true);
		this.add(via);
	}

	@Override
	public Component getTableCellRendererComponent(JTable table, Object value, boolean isSelected, boolean hasFocus, int row, int column)
	{
		StatusModel s = TweetDatabase.getStatus((Long) value);
		if ((table.getRowCount() - row) % 2 == 0)
		{
			setColor(Color.white);
		} else
		{
			setColor(new Color(224, 224, 224));
		}
		if (s.isFavorited())
		{
			setColor(new Color(255, 204, 51));
		}
		if (isSelected)
		{
			setColor(new Color(128, 255, 128));
		}

		for (Component c : this.getComponents())
		{
			c.setFont(new Font("Meiryo", Font.PLAIN, 12));
		}

		icon.setIcon(IconCache.getImage(s.getIconURL()));

		name.setText(s.getName());
		name.setLocation(48 + 10, 0);
		// getPreferredSizeの値、キャッシュする場合があるからヤクイ by @noko_k
		name.setSize(table.getSize().width - 48 - 10, Integer.MAX_VALUE);
		name.setSize(table.getSize().width - 48 - 10, name.getPreferredSize().height);
		// System.out.println(event.getText());
		// System.out.println(event.getPreferredSize());

		text.setText(s.getText());
		text.setLocation(48 + 10, name.getLocation().y + name.getPreferredSize().height);
		text.setSize(table.getSize().width - 48 - 10, Integer.MAX_VALUE);
		text.setSize(table.getSize().width - 48 - 10, text.getPreferredSize().height);
		// System.out.println(text.getPreferredSize());

		via.setText(s.getSource());
		via.setLocation(48 + 10, text.getLocation().y + text.getPreferredSize().height);
		via.setSize(table.getSize().width - 48 - 10, Integer.MAX_VALUE);
		via.setSize(table.getSize().width - 48 - 10, via.getPreferredSize().height);

		height = Math.max(58, name.getPreferredSize().height + text.getPreferredSize().height + via.getPreferredSize().height);
		this.setSize(0, height);

		//this.setPreferredSize(new Dimension(list.getSize().width - 48 - 10, Math.max(58, lineHeight + 1)));
		return this;
	}

	void setColor(Color c)
	{
		this.setBackground(c);
		icon.setBackground(c);
		name.setBackground(c);
		text.setBackground(c);
		via.setBackground(c);
	}

}

File: net/firsp/SwingBird/view/StatusRenderer.scala
package net.firsp.SwingBird.view

import javax.swing._
import java.awt.{Color, Component}
import net.firsp.SwingBird.model.StatusModel
import net.firsp.SwingBird.cache.{IconCache, Cache}

class StatusRenderer(table: JTable) extends StripedRendererBase(table) {
	val icon = new JLabel
	val name = new JTextArea
	val text = new JTextArea
	val via = new JTextArea

	setLayout(null)
	setOpaque(true)
	Array(name, text, via).foreach((t) => {
		t.setEditable(false)
		t.setLineWrap(true)
	})
	Array(icon, name, text, via).foreach(add)
	icon.setSize(48, 48)
	icon.setLocation(5, 5)

	def setColor(c: Color) = Array(this, icon, name, text, via).foreach(_.setBackground(c))

	override def getTableCellRendererComponent(table: JTable, value: Any, isSelected: Boolean, hasFocus: Boolean, row: Int, column: Int): Component = {
		super.getTableCellRendererComponent(table, value, isSelected, hasFocus, row, column)
		val m = value.asInstanceOf[StatusModel]
		val u = Cache.getOriginalUser(m)
		if (false) {
			//fav
			setColor(new Color(255, 204, 51))
		}
		if (m.rt) {
			setColor(new Color(135, 206, 235)) //skyblue
		}
		if (isSelected) {
			setColor(new Color(128, 255, 128))
		}
		icon.setIcon(IconCache.getImage(u.getProfileImageURL))

		val locateX = 48 + 10
		val sizeX = table.getSize().width - locateX
		val max = Integer.MAX_VALUE

		name.setText(u.getScreenName + " / " + u.getName)
		name.setLocation(locateX, 0)
		name.setSize(sizeX, max)
		name.setSize(sizeX, name.getPreferredSize.height)

		text.setText(m.text)
		text.setLocation(locateX, name.getLocation().y + name.getPreferredSize.height)
		text.setSize(sizeX, max)
		text.setSize(sizeX, text.getPreferredSize.height)

		via.setText(m.via)
		via.setLocation(locateX, text.getLocation().y + text.getPreferredSize.height)
		via.setSize(sizeX, max)
		via.setSize(sizeX, via.getPreferredSize.height)

		val height = Math.max(58, name.getPreferredSize.height + text.getPreferredSize.height + via.getPreferredSize.height)
		if (table.getRowHeight(row) != height) {
			table.setRowHeight(row, height)
		}
		this
	}
}

